### PR TITLE
Disable napari-console button when launched from jupyter

### LINKED ISSUE #687: Active napari viewer breaks ipywidgets events
## 🐛 Bug

If an instance of napari.Viewer() is present in a jupyter notebook, ipython widget events do not function properly.  The following error is sent to stderr:

`IPKernelApp] WARNING | No such comm: b161f0d81502401fab9931c48a91f974`

## To Reproduce

1. Run the following code in a notebook:

```python
%gui qt
```

Then:

```python
import napari
import ipywidgets as widgets
from IPython.display import display

viewer = napari.Viewer()

def print_it(ev):
    with outputter:
        if ev["name"] == "value":
            print(ev["new"])
            
select = widgets.Select(options=["image 1", "image 2"])
select.observe(print_it)
outputter = widgets.Output()
display(select, outputter)
```

## Expected behavior

Widget events should fire correctly.

## Environment

napari==0.2.5+2.g09a4117
PyQt5==5.13.2
jupyter-core==4.6.1
ipywidgets==7.5.1
ipython==7.9.0

### LINKED ISSUE #5211: Napari-console doesn't work when launching napari from jupyter
## 🐛 Bug

When launching napari from *ipython*, the terminal window is blank and is not responding (can't type).  However, it works just fine when launching napari from *python*. 

## To Reproduce

1. Create clean napari environment 

```
mamba create -n napari-clean python=3.10 napari --channel conda-forge --override-channels
```

2. Execute the example code below in ipython or jupyter notebook 

```
import napari
import numpy as np

arr = np.random.random((10,10,10))
viewer = napari.view_image(arr)
```

3. The terminal window in the napari program is blank and not responding. 

image

## Expected behavior

I expected that the terminal window in the napari looks like below screenshot, which I took when opening napari from python (not ipython).

image

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
```
napari: 0.4.16
Platform: macOS-12.6-x86_64-i386-64bit
System: MacOS 12.6
Python: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:43:44) [Clang 13.0.1 ]
Qt: 5.15.4
PyQt5: 5.15.7
NumPy: 1.23.3
SciPy: 1.9.1
Dask: 2022.9.2
VisPy: 0.10.0

OpenGL:
- GL version: 2.1 ATI-4.8.101
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 2.0

Plugins:
- console: 0.0.6
- napari-svg: 0.1.6
- scikit-image: 0.4.16
```

 - Any other relevant information:

### PR DESCRIPTION
**Pull request by psobolewskiPhD**
# Description
In https://github.com/napari/napari/pull/4240 the console button was disabled for ipython, but not jupyter. As a result, at present, if napari is launched from a jupyter notebook, e.g.:
```
import napari
viewer = napari.Viewer()
```
Then the console button is functional, but brings up a empty box—a non-functional console.
image

As the user is expected (?) to keep using the jupyter notebook, the console button should be disabled. So in this PR, I just add `in_jupyter` to the check introduced in https://github.com/napari/napari/pull/4240

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Fixes https://github.com/napari/napari/issues/5211
But sort of? From my look at the docs, the napari-console is not well documented to users, so need to make sure users arn't misunderstanding the console. I've been quite confused by this myself, plus https://github.com/napari/napari/issues/3338 is still a thing and it's **really** annoying.
I will note, that the `napari-console` is super powerful and a real game changer—every time I show it to someone it's like 🤯 

Maybe also closes https://github.com/napari/napari/issues/687

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change locally
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
I don't get whats up with the test at all—linux, pyside6? Seems ambitious...
😢

**Comment by Czaki** [comment]
> I don't get whats up with the test at all—linux, pyside6? Seems ambitious...

Just a typical problem with PySide6... 
From time to time, they do strange things with this. The version 6.4.0 of PySide6 and shiboken6  was released 5 hours ago. 

But it looks like they repeat an old error:

I think that there is a need to block it here https://github.com/napari/napari/blob/main/tox.ini#L78

@tlambert03 do you have any idea how to handle this?

**Review by goanpeca — APPROVED** [review]
Thanks for working on this @psobolewskiPhD ! 🚀

**Comment by andy-sweet** [comment]
TLDR: code changes look good and I think we should probably merge this.

> As the user is expected (?) to keep using the jupyter notebook, the console button should be disabled.

But I'm also a little unsure about the expected behavior here.

In the implementation of napari-console, we can [see that there is no support for connecting the napari console and an iPython shell launched from a terminal](https://github.com/napari/napari-console/blob/52eb3c480907df4c1118b7ed9920bf9c9a3948e7/napari_console/qt_console.py#L114).

But there is [an attempt to connect the napari console to an interactive shell launched from a jupyter notebook](https://github.com/napari/napari-console/blob/52eb3c480907df4c1118b7ed9920bf9c9a3948e7/napari_console/qt_console.py#L123).

On `main` (at cef660e74e7425a2b298262bd5695fa0927d36f2) on an Intel mac, this works - executing a command in either the notebook or in the napari console updates the shared state. I only tried a few basic operations though.

But if there are environments where this doesn't work (which seems to be the case), then I think we should probably disable this as you've done here.

More generally, my feeling is that if you're launching napari from iPython or a jupyter notebook (or some other interactive shell or repl), you probably don't want the napari console. I think is especially true for a notebook, where you might expect that the notebook maintains some kind of record of the executed code. But I'd be curious about usage and opinions from people who are actually using napari from notebooks.

**Comment by andy-sweet** [comment]
> I don't get whats up with the test at all—linux, pyside6? Seems ambitious...

I think if you merge `main` and push, that test should pass now after #5244 was merged.

**Comment by psobolewskiPhD** [comment]
> On `main` (at [cef660e](https://github.com/napari/napari/commit/cef660e74e7425a2b298262bd5695fa0927d36f2)) on an Intel mac, this works - executing a command in either the notebook or in the napari console updates the shared state. I only tried a few basic operations though.

Interesting, the OP of the issue https://github.com/napari/napari/issues/5211 also has x86 macOS 12.6
On my arm64 I can reproduce the issue with `jupyter qtconsole` as well (with freshly pulled `main`):
```
Jupyter QtConsole 5.3.2
Python 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:22) [Clang 13.0.1 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.5.0 -- An enhanced Interactive Python. Type '?' for help.

import napari
viewer = napari.Viewer()
```
image

We're both running python 3.10 though. Maybe it's that? let me check.

**Comment by psobolewskiPhD** [comment]
On python 3.9 I get an error when opening the console after launching from jupyter qtconsole.
```
ERROR:tornado.general:Uncaught exception in ZMQStream callback
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 557, in _run_callback
    callback(*args, **kwargs)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/jupyter_client/threaded.py", line 123, in _handle_recv
    msg_list = self.ioloop._asyncio_event_loop.run_until_complete(get_msg(future_msg))
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 623, in run_until_complete
    self._check_running()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 585, in _check_running
    raise RuntimeError(
RuntimeError: Cannot run the event loop while another loop is running
ERROR:tornado.general:Uncaught exception in zmqstream callback
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 578, in _handle_events
    self._handle_recv()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 607, in _handle_recv
    self._run_callback(callback, msg)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 557, in _run_callback
    callback(*args, **kwargs)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/jupyter_client/threaded.py", line 123, in _handle_recv
    msg_list = self.ioloop._asyncio_event_loop.run_until_complete(get_msg(future_msg))
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 623, in run_until_complete
    self._check_running()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 585, in _check_running
    raise RuntimeError(
RuntimeError: Cannot run the event loop while another loop is running
ERROR:tornado.general:Uncaught exception in ZMQStream callback
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 557, in _run_callback
    callback(*args, **kwargs)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/jupyter_client/threaded.py", line 123, in _handle_recv
    msg_list = self.ioloop._asyncio_event_loop.run_until_complete(get_msg(future_msg))
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 623, in run_until_complete
    self._check_running()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 585, in _check_running
    raise RuntimeError(
RuntimeError: Cannot run the event loop while another loop is running
ERROR:tornado.general:Uncaught exception in zmqstream callback
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 578, in _handle_events
    self._handle_recv()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 607, in _handle_recv
    self._run_callback(callback, msg)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/zmq/eventloop/zmqstream.py", line 557, in _run_callback
    callback(*args, **kwargs)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/site-packages/jupyter_client/threaded.py", line 123, in _handle_recv
    msg_list = self.ioloop._asyncio_event_loop.run_until_complete(get_msg(future_msg))
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 623, in run_until_complete
    self._check_running()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/apple-TF210/lib/python3.9/asyncio/base_events.py", line 585, in _check_running
    raise RuntimeError(
RuntimeError: Cannot run the event loop while another loop is running
```

No errors if using a jupyter notebook, but same effect of a blank console.

**Comment by psobolewskiPhD** [comment]
> > I don't get whats up with the test at all—linux, pyside6? Seems ambitious...
> 
> I think if you merge `main` and push, that test should pass now after #5244 was merged.

@andy-sweet Sorry to be ignorant, but just want to be sure:
to do this I pull main, then checkout my PR branch, then `git merge main` and that's it? Or is this where I need to rebase? Still getting the hang of all this.
❤️ 

Edit: @Czaki helped me figure it out!

**Comment by andy-sweet** [comment]
> @andy-sweet Sorry to be ignorant, but just want to be sure:
> to do this I pull main, then checkout my PR branch, then `git merge main` and that's it? Or is this where I need to rebase? Still getting the hang of all this.

That's what I do.

The merge and rebase commands are two different ways of integrating changes from one branch to another. Merge only adds commits (including special merge commits) to a branch, while rebase rewrites commits. I always find the Atlassian git guide to be pretty great and [it covers this topic too](https://www.atlassian.com/git/tutorials/merging-vs-rebasing).

When using github's squash and merge, I generally don't really find rebase useful and I'm not bothered by merge commits on my fork's branches, so I tend to avoid it. But there likely are some cases when it would be appropriate. And I think some people generally prefer it.

**Comment by psobolewskiPhD** [comment]
Holy cow, that was really helpful. ❤️

**Review by andy-sweet — APPROVED** [review]

**Review by goanpeca — APPROVED** [review]

**Comment by andy-sweet** [comment]
@psobolewskiPhD : feel free to merge this when you want.

There might be two follow-up issues worth writing up, both outside of this repo.

1. Clarify the intended behavior (i.e. the console is disabled when launching from iPython or a Jupyter notebook) in the napari.org docs. The [getting started](https://napari.org/stable/tutorials/fundamentals/getting_started.html#jupyter-notebook-usage) page seems like a good place to do this.
2. Consider [removing the connection to the jupyter notebook Python kernel in the napari-console plugin](https://github.com/napari/napari-console/blob/52eb3c480907df4c1118b7ed9920bf9c9a3948e7/napari_console/qt_console.py#L123). I don't entirely understand this code, but I suspect the intent is to allow the console in napari to share the Python kernel started by Jupyter so that definitions made in the console can be used in the notebook (which doesn't seem particularly useful to me).
